### DISCUSSION TITLE
Gossipsub forwards messages in fully connected small mesh

### DISCUSSION THREAD

**Discussion by RustMania** [libp2p/rust-libp2p#3127] [Q&A]
Recently I observed that network packets in gossipsub are not delivered occasionally. This happens mostly on slow VMs with low CPU resource.  My application uses a protocol , where most of messages are broadcast to small fixed amount of parties, e.g 3-4. The protocol stopped occasionally after a timeout, because some of broadcasts did not seem to reach the destination.
 
After close examination of inputs,  I found that a node subscribed to a topic receives _all_ messages from that topic, however sometimes Gossipsub forwards messages via another node, even in very small fully connected mesh. As a result, the field **propagation_source** in GossipsubEvent::Message will be set to the peer id of a forwarder. Therefore the field is useless to identify  a message sender. 
To fix the problem I looked into the  GossipsubMessage, where I found the field **source**, which is Option<PeerId>, but  it seems to be set all the time to correct source id, at least for now and in my case.
 
I guess the forwarding is a feature of Gossipsub architecture aimed to build sparse meshes, so the field in the Event enum is just  an identity of the forwarder. If message is forwarded and  the field in GossipsubMessage is set to None, it leaves receiver w/o chance to identify a sender. _Under what condition the field becomes None?_

Finally, the floodsub protocol must be significantly simpler than gossipsub , just guessing from the implementation. 
_Can the entire problem  with the identification of the source be avoided by using floodsub instead?_

**Comment by thomaseizinger** [comment]
> _Under what condition the field becomes None?_

See this snippet:

https://github.com/libp2p/rust-libp2p/blob/c32f03c3170a8123cb480ece9aa511a6f9ea6dd5/protocols/gossipsub/src/behaviour.rs#L2762-L2850

The sender can choose what to set this field to. In an untrusted environment, you can only trust it when it comes with a signature as well.

If you have a small mesh, then yes, `floodsub` might be a better choice. It has really good latency properties but will flood all connected peers with every incoming message (except the sender).
